File: haskell/lib/Tidepool/Actor.hs
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE FlexibleContexts #-}
{-# LANGUAGE GADTs #-}
{-# LANGUAGE KindSignatures #-}
{-# LANGUAGE PatternSynonyms #-}
{-# LANGUAGE RankNTypes #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE TypeApplications #-}
{-# LANGUAGE TypeOperators #-}

-- | Typed observation of exact actor incarnations.
--
-- Copies of an 'ActorRef' observe the same immutable terminal result.
-- Successful @exit@ values may be arbitrary Haskell values, including
-- closures and lazy structures; 'awaitExit' is repeatable and never
-- serializes them.
--
-- Target failure and cancellation are domain-visible lifecycle outcomes.
-- An invalid reference, a machine-boundary violation, or cancellation of the
-- waiter is a runtime failure and does not fabricate an 'ActorExit'.
module Tidepool.Actor
  ( ActorRef
  , ActorDefinition
  , pattern ActorDefinition
  , label
  , effectProfile
  , initialization
  , behavior
  , onShutdown
  , Source
  , progressSource
  , settlementSource
  , withSources
  , stateful
  , EffectProfile (..)
  , LaunchRole (..)
  , ReadOnlyEffects
  , ReadWriteEffects
  , ShutdownReason (..)
  , startActor
  , beginActorForkGroup
  , startActorFork
  , commitActorForkGroup
  , abortActorForkGroup
  , runActor
  , call
  , cast
  , drainActor
  , receive
  , serve
  , ActorExit (..)
  , ActorFailure (..)
  , CancelReason (..)
  , awaitExit
  , pollExit
  ) where

import Control.Monad.Freer (Eff, Member, raise, send)
import Data.Text (Text)
import Prelude
import Tidepool.Actor.Source (Source, progressSource, settlementSource, installSource)

import Tidepool.Actor.Internal
  ( ActorDefinition
  , pattern ActorDefinition
  , label
  , effectProfile
  , initialization
  , behavior
  , onShutdown
  , actorLaunchWorktrees
  , actorSources
  , withSources
  , ActorRef (..)
  , EffectProfile (..)
  , ReadOnlyEffects
  , ReadWriteEffects
  , ShutdownReason (..)
  )
import Tidepool.Effects.Core
  ( Actor (..)
  , ActorLaunchRole (..)
  , ActorEffectProfile (..)
  , ActorKernel (..)
  , ActorLocal (..)
  , ActorTerminalStatus (..)
  )
import Tidepool.Internal.ExitCell
  ( fillExitCell
  , newExitCell
  , readExitCell
  )

data LaunchRole
  = RootRole
  | ResearchRole
  | CodingRole
  | ScaffoldingRole
  | IntegrationRole
  | InheritedRole
  deriving (Show, Eq)

-- | Failure reported for a terminated actor.
newtype ActorFailure = ActorFailure { actorFailureSummary :: Text }
  deriving (Show, Eq)

-- | Why an actor was cancelled rather than completing or failing.
newtype CancelReason = CancelReason { cancelReasonSummary :: Text }
  deriving (Show, Eq)

-- | The immutable terminal result retained by one exact 'ActorRef'.
data ActorExit exit
  = Completed exit
  | Failed ActorFailure
  | Cancelled CancelReason
  deriving (Show, Eq)

-- | Start a fresh actor from an authored definition. Tidepool captures the
-- definition's exact Haskell environment internally; authored code does not
-- manage a separate deployment value or preparation step.
--
-- This operation returns only after the child has installed its behavior and
-- reached readiness.
{-# NOINLINE startActor #-}
startActor
  :: forall effs startup api exit
   . Member Actor effs
  => ActorDefinition startup api exit
  -> startup
  -> Eff effs (ActorRef api exit)
startActor definition@ActorDefinition
  { label = actorLabel
  , effectProfile = profile
  , initialization = startupAction
  , behavior = install
  , onShutdown = shutdownAction
  } startup = do
  let cell = newExitCell startup
      shutdownEntry reasonCode =
        raiseKernel (shutdownAction (decodeShutdownReason reasonCode))
      entry _ = do
        send (ActorInstallShutdownWith 0 shutdownEntry)
        initial <- raiseKernel (startupAction startup)
        mapM_ installSource (actorSources definition)
        send ActorReadyWith
        result <- raiseKernel (install startup initial)
        case fillExitCell cell result of
          () -> pure ()
  (actorId, incarnation, _) <- send
    (ActorStartWith actorLabel entry ActorInheritedRole (profileCode profile) (actorLaunchWorktrees definition))
  pure (ActorRef actorId incarnation cell)

-- | Trusted context-fork launch. The runtime snapshots the caller's lexical
-- environment and the host forks its provider conversation at the active
-- tool call. Normal authored Shoal code reaches this through `unfold`.
{-# NOINLINE startActorFork #-}
beginActorForkGroup
  :: Member Actor effs
  => Bool
  -> Text
  -> [Text]
  -> Eff effs (Int, Text, [Text])
beginActorForkGroup relative group branches =
  send (ActorBeginForkGroupWith relative group branches)

startActorFork
  :: forall effs startup api exit
  . Member Actor effs
  => LaunchRole
  -> Int
  -> ActorDefinition startup api exit
  -> startup
  -> Eff effs (ActorRef api exit, Text)
startActorFork launchRole forkGroup definition@ActorDefinition
  { label = actorLabel
  , effectProfile = profile
  , initialization = startupAction
  , behavior = install
  , onShutdown = shutdownAction
  } startup = do
  let cell = newExitCell startup
      shutdownEntry reasonCode =
        raiseKernel (shutdownAction (decodeShutdownReason reasonCode))
      entry _ = do
        send (ActorInstallShutdownWith 0 shutdownEntry)
        initial <- raiseKernel (startupAction startup)
        mapM_ installSource (actorSources definition)
        send ActorReadyWith
        result <- raiseKernel (install startup initial)
        case fillExitCell cell result of
          () -> pure ()
  (actorId, incarnation, allocatedPath) <- send
    (ActorForkWith actorLabel entry forkGroup (roleCode launchRole) (profileCode profile) (actorLaunchWorktrees definition))
  pure (ActorRef actorId incarnation cell, allocatedPath)

commitActorForkGroup :: Member Actor effs => Int -> Eff effs ()
commitActorForkGroup = send . ActorCommitForkGroupWith

abortActorForkGroup :: Member Actor effs => Int -> Eff effs ()
abortActorForkGroup = send . ActorAbortForkGroupWith

roleCode :: LaunchRole -> ActorLaunchRole
roleCode RootRole = ActorRootRole
roleCode ResearchRole = ActorResearchRole
roleCode CodingRole = ActorCodingRole
roleCode ScaffoldingRole = ActorScaffoldingRole
roleCode IntegrationRole = ActorIntegrationRole
roleCode InheritedRole = ActorInheritedRole

profileCode :: EffectProfile protocol effs -> ActorEffectProfile
profileCode ReadWrite = ActorReadWriteProfile
profileCode ReadOnly = ActorReadOnlyProfile

decodeShutdownReason :: Int -> ShutdownReason
decodeShutdownReason 0 = ShutdownCompleted
decodeShutdownReason 1 = ShutdownFailed
decodeShutdownReason _ = ShutdownCancelled

raiseKernel :: Eff effs a -> Eff (ActorKernel ': effs) a
raiseKernel = raise

-- | Start a supervised one-shot actor and wait for its exact terminal value.
runActor
  :: Member Actor effs
  => ActorDefinition startup api exit
  -> startup
  -> Eff effs (ActorExit exit)
runActor definition startup = startActor definition startup >>= awaitExit

-- | Make one synchronous request to an exact actor incarnation. Runtime
-- lifecycle failure abandons this fragment; it is never fabricated as a
-- value of the protocol's result type.
call
  :: Member Actor effs
  => ActorRef protocol exit
  -> protocol result
  -> Eff effs result
call (ActorRef actorId incarnation _) request =
  send (ActorCallWith (actorId, incarnation) request)

-- | Transfer one unit-result request into an exact actor mailbox. Returning
-- means the mailbox accepted ownership, not that the target handled it.
cast
  :: Member Actor effs
  => ActorRef protocol exit
  -> protocol ()
  -> Eff effs ()
cast (ActorRef actorId incarnation _) request =
  send (ActorCastWith (actorId, incarnation) request)

-- | Close an owned stateful actor's sources and mailbox and request completion
-- of accepted messages. Returns after admission closes; 'awaitExit' observes
-- the final state. A paused handler needs repair before draining.
drainActor
  :: Member Actor effs
  => ActorRef protocol state
  -> Eff effs ()
drainActor (ActorRef actorId incarnation _) =
  send (ActorDrainWith (actorId, incarnation))

-- | Handle one request of any result index and return the next actor state.
-- Reply authority remains in the runtime; authored code returns an ordinary
-- pair and never receives a token it could duplicate or forget.
{-# OPAQUE receive #-}
receive
  :: forall next protocol effs
   . Member (ActorLocal protocol) effs
  => (forall result. protocol result -> Eff effs (result, next))
  -> Eff effs next
receive = receiveSited @next 0

-- Extractor substrate. The stable site key correlates the receive suspension
-- with its two private settlement steps and gives observability a source-level
-- identity without exposing a reply token.
{-# OPAQUE receiveSited #-}
receiveSited
  :: forall next protocol effs
   . Member (ActorLocal protocol) effs
  => Int
  -> (forall result. protocol result -> Eff effs (result, next))
  -> Eff effs next
receiveSited site handler = send (ActorReceiveWith site kernelHandler)
  where
    kernelHandler :: forall result. protocol result -> Eff (ActorKernel ': effs) ()
    kernelHandler request = do
      (reply, next) <- raiseKernel (handler request)
      send (ActorReplyWith site reply)
      send (ActorContinueWith site next)

-- | Serve requests forever with explicit Haskell state. Actors that may exit
-- in response to a message use 'receive' directly and return normally.
serve
  :: forall state protocol effs exit
   . Member (ActorLocal protocol) effs
  => state
  -> (forall result. state -> protocol result -> Eff effs (result, state))
  -> Eff effs exit
{-# OPAQUE serve #-}
serve state step = serveSited @state unreachableSiteId state step
  where
    unreachableSiteId = error
      "serve: unreachable — extract must assign its receive site at the fully applied call"

-- A recursive server owns one stable receive site. The extractor rewrites the
-- public, fully-applied `serve` call where `state` is concrete; recursion then
-- reuses that site instead of trying to extract this polymorphic library body.
{-# OPAQUE serveSited #-}
serveSited
  :: forall state protocol effs exit
   . Member (ActorLocal protocol) effs
  => Int
  -> state
  -> (forall result. state -> protocol result -> Eff effs (result, state))
  -> Eff effs exit
serveSited site state step =
  receiveSited @state site (step state) >>= \next -> serveSited site next step

-- | Define a mailbox actor with retained state. Each successful message commits
-- its reply and next state together; a failed handler pauses for replacement.
stateful
  :: Member (ActorLocal protocol) effs
  => Text
  -> EffectProfile protocol effs
  -> (forall result. state -> protocol result -> Eff effs (result, state))
  -> ActorDefinition state protocol state
stateful actorLabel profile step = ActorDefinition
  { label = actorLabel
  , effectProfile = profile
  , initialization = pure
  , behavior = \_ -> statefulLoop step
  , onShutdown = const (pure ())
  }

statefulLoop
  :: forall state protocol effs
   . Member (ActorLocal protocol) effs
  => (forall result. state -> protocol result -> Eff effs (result, state))
  -> state
  -> Eff effs state
statefulLoop step state = do
  send @(ActorLocal protocol) (ActorCheckpointWith 0 state)
  next <- receiveSited @(Maybe state) 0 (\message -> do
    (reply, nextState) <- step state message
    pure (reply, Just nextState))
  case next of
    Nothing -> pure state
    Just nextState -> statefulLoop step nextState

-- | Observe this exact actor incarnation's retained terminal result.
--
-- Completion is published into the reference's cell before the Rust terminal
-- transition. Seeing completion with an empty cell is therefore an engine
-- invariant violation, not a lifecycle case authors must handle.
awaitExit :: Member Actor effs => ActorRef api exit -> Eff effs (ActorExit exit)
awaitExit (ActorRef actorId incarnation cell) = do
  terminal <- send (ActorWaitWith (actorId, incarnation))
  case terminal of
    ActorCompletedStatus ->
      case readExitCell terminal cell of
        Just value -> pure (Completed value)
        Nothing -> error "Tidepool.Actor.awaitExit: completed actor has an empty exit cell"
    ActorFailedStatus summary -> pure (Failed (ActorFailure summary))
    ActorCancelledStatus summary -> pure (Cancelled (CancelReason summary))

-- | Observe an exact actor without parking the caller. Completion reads the
-- same shared exit cell as 'awaitExit'; repeated polls therefore return the
-- same typed result and never consume custody.
pollExit :: Member Actor effs => ActorRef api exit -> Eff effs (Maybe (ActorExit exit))
pollExit (ActorRef actorId incarnation cell) = do
  terminal <- send (ActorPollWith (actorId, incarnation))
  pure (terminal >>= decodeTerminal cell)
  where
    decodeTerminal retained status = case status of
      ActorCompletedStatus ->
        case readExitCell status retained of
          Just value -> Just (Completed value)
          Nothing -> error "Tidepool.Actor.pollExit: completed actor has an empty exit cell"
      ActorFailedStatus summary -> Just (Failed (ActorFailure summary))
      ActorCancelledStatus summary -> Just (Cancelled (CancelReason summary))

File: tidepool-protocol/src/effects/actor.rs
//! The actor-runtime suspension boundary.
//!
//! The public Haskell API lives in `Tidepool.Actor`; this schema owns only the
//! outward runtime requests. A start carries one
//! existentially row-typed child entry as its field-1 live payload. A wait
//! carries an exact Rust routing identity and returns terminal metadata. The
//! successful exit value never crosses either request: it remains in the
//! managed Haskell cell carried by the corresponding `ActorRef`.
//!
//! There is no `tidepool-handlers` handler.  The request is decoded and
//! serviced by `tidepool-actor`, whose registry owns exact-incarnation wait
//! semantics.

use crate::hs::HsType;
use crate::schema::{
    Arg, Effect, HandlingClass, JsonInstance, Polymorphism, RustBinding, SumVariant, TypeDef,
    TypeShape, Verb, WireDerive, WireDerives,
};
fn address_type() -> HsType {
    HsType::Tuple(vec![HsType::Int, HsType::Int])
}

fn launched_actor_type() -> HsType {
    HsType::Tuple(vec![HsType::Int, HsType::Int, HsType::Text])
}

/// The actor effect's internal start and wait requests.
#[must_use]
pub fn actor() -> Effect {
    Effect {
        name: "Actor",
        authored_surface: crate::schema::AuthoredSurface::OPAQUE,
        handler: "ActorDecodeHandler",
        handler_module: "actor",
        req_enum: "ActorReq",
        decl_fn: "actor_decl",
        description: &[
            "Typed actor lifecycle and communication. Authored code uses `Tidepool.Actor`; ",
            "the constructors in this effect are runtime substrate, not a second public API.",
        ],
        prompt_card: Some(&[
            "`Tidepool.Actor` provides exact-incarnation actor references and typed exit ",
            "observation. `awaitExit ref` returns `Completed value`, `Failed reason`, or ",
            "`Cancelled reason`; it is repeatable and preserves closure-valued exits.",
        ]),
        type_params: &[],
        default_row_args: &[],
        helpers_row_polymorphic: true,
        extra_imports: &["import Tidepool.Actor"],
        type_defs: vec![
            TypeDef {
                name: "ActorLaunchRole",
                wire_rust: None,
                shape: TypeShape::Sum {
                    variants: vec![
                        SumVariant {
                            ctor: "ActorRootRole",
                            fields: positional_fields![],
                            doc: &[],
                        },
                        SumVariant {
                            ctor: "ActorResearchRole",
                            fields: positional_fields![],
                            doc: &[],
                        },
                        SumVariant {
                            ctor: "ActorCodingRole",
                            fields: positional_fields![],
                            doc: &[],
                        },
                        SumVariant {
                            ctor: "ActorScaffoldingRole",
                            fields: positional_fields![],
                            doc: &[],
                        },
                        SumVariant {
                            ctor: "ActorIntegrationRole",
                            fields: positional_fields![],
                            doc: &[],
                        },
                        SumVariant {
                            ctor: "ActorInheritedRole",
                            fields: positional_fields![],
                            doc: &[],
                        },
                    ],
                },
                json: JsonInstance::None,
                derives: WireDerives(&[
                    WireDerive::Debug,
                    WireDerive::Clone,
                    WireDerive::PartialEq,
                    WireDerive::Eq,
                ]),
                domain: None,
                doc: &["Semantic role selected before runtime policy projection."],
            },
            TypeDef {
                name: "ActorEffectProfile",
                wire_rust: None,
                shape: TypeShape::Sum {
                    variants: vec![
                        SumVariant {
                            ctor: "ActorReadWriteProfile",
                            fields: positional_fields![],
                            doc: &[],
                        },
                        SumVariant {
                            ctor: "ActorReadOnlyProfile",
                            fields: positional_fields![],
                            doc: &[],
                        },
                    ],
                },
                json: JsonInstance::None,
                derives: WireDerives(&[
                    WireDerive::Debug,
                    WireDerive::Clone,
                    WireDerive::PartialEq,
                    WireDerive::Eq,
                ]),
                domain: None,
                doc: &["Closed interpreter profile selected for one actor incarnation."],
            },
            TypeDef {
                name: "ActorTerminalStatus",
                wire_rust: None,
                shape: TypeShape::Sum {
                    variants: vec![
                        SumVariant {
                            ctor: "ActorCompletedStatus",
                            fields: positional_fields![],
                            doc: &[],
                        },
                        SumVariant {
                            ctor: "ActorFailedStatus",
                            fields: positional_fields![HsType::Text],
                            doc: &[],
                        },
                        SumVariant {
                            ctor: "ActorCancelledStatus",
                            fields: positional_fields![HsType::Text],
                            doc: &[],
                        },
                    ],
                },
                json: JsonInstance::None,
                derives: WireDerives(&[
                    WireDerive::Debug,
                    WireDerive::Clone,
                    WireDerive::PartialEq,
                    WireDerive::Eq,
                ]),
                domain: None,
                doc: &[],
            },
            TypeDef {
                name: "ActorCallStatus",
                wire_rust: None,
                shape: TypeShape::Sum {
                    variants: vec![
                        SumVariant {
                            ctor: "ActorCallSucceeded",
                            fields: positional_fields![],
                            doc: &[],
                        },
                        SumVariant {
                            ctor: "ActorCallFailed",
                            fields: positional_fields![HsType::Text],
                            doc: &[],
                        },
                    ],
                },
                json: JsonInstance::None,
                derives: WireDerives(&[
                    WireDerive::Debug,
                    WireDerive::Clone,
                    WireDerive::PartialEq,
                    WireDerive::Eq,
                ]),
                domain: None,
                doc: &["Result of a unit-returning actor call whose lifecycle failure is data."],
            },
        ],
        foreign_types: &[],
        errors: None,
        verbs: vec![
            Verb {
                ctor: "ActorBeginForkGroupWith",
                method: "actor_begin_fork_group_with",
                args: vec![
                    Arg {
                        name: "relative",
                        ty: HsType::Bool,
                        rust: RustBinding::Derived,
                    },
                    Arg {
                        name: "group",
                        ty: HsType::Text,
                        rust: RustBinding::Derived,
                    },
                    Arg {
                        name: "branches",
                        ty: HsType::List(Box::new(HsType::Text)),
                        rust: RustBinding::Derived,
                    },
                ],
                ret: HsType::Tuple(vec![
                    HsType::Int,
                    HsType::Text,
                    HsType::List(Box::new(HsType::Text)),
                ]),
                errors: None,
                handling: HandlingClass::Actor,
                extract: None,
            },
            Verb {
                ctor: "ActorStartWith",
                method: "actor_start_with",
                args: vec![
                    Arg {
                        name: "label",
                        ty: HsType::Text,
                        rust: RustBinding::Derived,
                    },
                    Arg {
                        name: "entry",
                        ty: HsType::func(
                            HsType::Int,
                            HsType::app(
                                HsType::app(HsType::Named("Eff"), HsType::Var("childEffs")),
                                HsType::Unit,
                            ),
                        ),
                        rust: RustBinding::CoreValue,
                    },
                    Arg {
                        name: "role",
                        ty: HsType::Named("ActorLaunchRole"),
                        rust: RustBinding::Path("crate::ActorLaunchRoleWire"),
                    },
                    Arg {
                        name: "profile",
                        ty: HsType::Named("ActorEffectProfile"),
                        rust: RustBinding::Path("crate::ActorEffectProfileWire"),
                    },
                    Arg {
                        name: "launchWorktrees",
                        ty: HsType::List(Box::new(HsType::Text)),
                        rust: RustBinding::Derived,
                    },
                ],
                ret: launched_actor_type(),
                errors: None,
                handling: HandlingClass::Actor,
                extract: None,
            },
            Verb {
                ctor: "ActorForkWith",
                method: "actor_fork_with",
                args: vec![
                    Arg {
                        name: "label",
                        ty: HsType::Text,
                        rust: RustBinding::Derived,
                    },
                    Arg {
                        name: "entry",
                        ty: HsType::func(
                            HsType::Int,
                            HsType::app(
                                HsType::app(HsType::Named("Eff"), HsType::Var("childEffs")),
                                HsType::Unit,
                            ),
                        ),
                        rust: RustBinding::CoreValue,
                    },
                    Arg {
                        name: "forkGroup",
                        ty: HsType::Int,
                        rust: RustBinding::Derived,
                    },
                    Arg {
                        name: "role",
                        ty: HsType::Named("ActorLaunchRole"),
                        rust: RustBinding::Path("crate::ActorLaunchRoleWire"),
                    },
                    Arg {
                        name: "profile",
                        ty: HsType::Named("ActorEffectProfile"),
                        rust: RustBinding::Path("crate::ActorEffectProfileWire"),
                    },
                    Arg {
                        name: "launchWorktrees",
                        ty: HsType::List(Box::new(HsType::Text)),
                        rust: RustBinding::Derived,
                    },
                ],
                ret: launched_actor_type(),
                errors: None,
                handling: HandlingClass::Actor,
                extract: None,
            },
            Verb {
                ctor: "ActorCommitForkGroupWith",
                method: "actor_commit_fork_group_with",
                args: vec![Arg {
                    name: "forkGroup",
                    ty: HsType::Int,
                    rust: RustBinding::Derived,
                }],
                ret: HsType::Unit,
                errors: None,
                handling: HandlingClass::Actor,
                extract: None,
            },
            Verb {
                ctor: "ActorAbortForkGroupWith",
                method: "actor_abort_fork_group_with",
                args: vec![Arg {
                    name: "forkGroup",
                    ty: HsType::Int,
                    rust: RustBinding::Derived,
                }],
                ret: HsType::Unit,
                errors: None,
                handling: HandlingClass::Actor,
                extract: None,
            },
            Verb {
                ctor: "ActorWaitWith",
                method: "actor_wait_with",
                args: vec![Arg {
                    name: "actor",
                    ty: address_type(),
                    rust: RustBinding::Path("(i64, i64)"),
                }],
                ret: HsType::Named("ActorTerminalStatus"),
                errors: None,
                handling: HandlingClass::Actor,
                extract: None,
            },
            Verb {
                ctor: "ActorPollWith",
                method: "actor_poll_with",
                args: vec![Arg {
                    name: "actor",
                    ty: address_type(),
                    rust: RustBinding::Path("(i64, i64)"),
                }],
                ret: HsType::maybe(HsType::Named("ActorTerminalStatus")),
                errors: None,
                handling: HandlingClass::Actor,
                extract: None,
            },
            Verb {
                ctor: "ActorCallWith",
                method: "actor_call_with",
                args: vec![
                    Arg {
                        name: "actor",
                        ty: address_type(),
                        rust: RustBinding::Path("(i64, i64)"),
                    },
                    Arg {
                        name: "request",
                        ty: HsType::app(HsType::Var("protocol"), HsType::Var("result")),
                        rust: RustBinding::CoreValue,
                    },
                ],
                ret: HsType::Var("result"),
                errors: None,
                handling: HandlingClass::Actor,
                extract: None,
            },
            Verb {
                ctor: "ActorTryCallWith",
                method: "actor_try_call_with",
                args: vec![
                    Arg {
                        name: "actor",
                        ty: address_type(),
                        rust: RustBinding::Path("(i64, i64)"),
                    },
                    Arg {
                        name: "request",
                        ty: HsType::app(HsType::Var("protocol"), HsType::Unit),
                        rust: RustBinding::CoreValue,
                    },
                ],
                ret: HsType::Named("ActorCallStatus"),
                errors: None,
                handling: HandlingClass::Actor,
                extract: None,
            },
            Verb {
                ctor: "ActorCastWith",
                method: "actor_cast_with",
                args: vec![
                    Arg {
                        name: "actor",
                        ty: address_type(),
                        rust: RustBinding::Path("(i64, i64)"),
                    },
                    Arg {
                        name: "request",
                        ty: HsType::app(HsType::Var("protocol"), HsType::Unit),
                        rust: RustBinding::CoreValue,
                    },
                ],
                ret: HsType::Unit,
                errors: None,
                handling: HandlingClass::Actor,
                extract: None,
            },
            Verb {
                ctor: "ActorDrainWith",
                method: "actor_drain_with",
                args: vec![Arg {
                    name: "actor",
                    ty: address_type(),
                    rust: RustBinding::Path("(i64, i64)"),
                }],
                ret: HsType::Unit,
                errors: None,
                handling: HandlingClass::Actor,
                extract: None,
            },
        ],
        // The public wrapper needs the managed cell carried by `ActorRef`, so
        // it is authored in Tidepool.Actor rather than emitted as a second,
        // raw helper here.
        helpers: Vec::new(),
        polymorphism: Polymorphism::None,
        dispatched: false,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn actor_entries_remain_the_field_one_live_payload() {
        let effect = actor();
        for constructor in ["ActorStartWith", "ActorForkWith"] {
            let verb = effect
                .verbs
                .iter()
                .find(|verb| verb.ctor == constructor)
                .unwrap();
            assert!(matches!(verb.args[1].rust, RustBinding::CoreValue));
            assert_eq!(
                verb.args
                    .iter()
                    .filter(|argument| matches!(argument.rust, RustBinding::CoreValue))
                    .count(),
                1
            );
        }
    }
}
